### PR TITLE
PyTorch Temporal memory

### PR DESCRIPTION
**Pull request by abhi-iyer**
PyTorch version of TM. Comes with examples and test cases. 

Two extensions of TM: sequence memory and pair memory.

@subutai @heikohoff

### CODE REVIEW

File: src/nupic/research/frameworks/htm/temporal_memory/pair_memory_apical_tiebreak.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+from nupic.research.frameworks.htm.temporal_memory import TemporalMemoryApicalTiebreak
+
+import torch
+
+real_type = torch.float32
+int_type = torch.int64
```
Comment (lucasosouza) [inline src/nupic/research/frameworks/htm/temporal_memory/pair_memory_apical_tiebreak.py:27]: int64 is a really wide range, do we really use all that.

File: src/nupic/research/frameworks/htm/temporal_memory/sequence_memory_apical_tiebreak.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+            "permanence_decrement" : permanence_decrement,
+            "basal_segment_incorrect_decrement" : basal_segment_incorrect_decrement,
+            "apical_segment_incorrect_decrement" : apical_segment_incorrect_decrement,
+            "max_synapses_per_segment" : max_synapses_per_segment,
+            "seed" : seed
+        }
+
+        super().__init__(**params)
+
+        self.previous_apical_input = torch.empty(0, dtype=int_type).to(device)
```
Comment (lucasosouza) [inline src/nupic/research/frameworks/htm/temporal_memory/sequence_memory_apical_tiebreak.py:76]: If you are initializing a tensor, you can instead pass a device argument to the `empty` function. Any pytorch function that initializes a tensor will take both a dtype and a device argument.

File: src/nupic/research/frameworks/htm/temporal_memory/sequence_memory_apical_tiebreak.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+        apical_input=torch.Tensor().to(device),
+        apical_growth_candidates=None,
+        learn=True
+    ):
+        """
+        perform one timestep: activate the specified minicolumn using the predictions
+        from the previous timestep and then learn. form a new set of predictions
+        using the newly active cells and the apical input.
+
+        `active_minicolumns` (torch.Tensor) contains the active minicolumns.
```
Comment (lucasosouza) [inline src/nupic/research/frameworks/htm/temporal_memory/sequence_memory_apical_tiebreak.py:106]: This looks good, if it becomes stable code we might want to standardize the docstring style to a single format. I am not a big fan of the style we current use in nupic.research, I personally prefer numpy style, but I agree that standardizing is a good idea.

**Review by lucasosouza — APPROVED** [review]
Thanks for including me. Did a very shallow review with not very relevant comments, I am not familiar with this codebase and Luiz is probably a better person to review it. 

Some of the functions are long and multipurposed, so hard to read and maintain. If this becomes stable code we might want to break down a few of these into more logical steps (following the single responsability principle) and maybe have a diagram in a Readme or docstrings detailing the training/evaluation loop flow.

File: src/nupic/research/frameworks/htm/temporal_memory/pair_memory_apical_tiebreak.py (hunk 1 above)
Comment (abhi-iyer) [reply src/nupic/research/frameworks/htm/temporal_memory/pair_memory_apical_tiebreak.py:27]: int64 is the minimum width needed for Tensors that index into other Tensors

File: src/nupic/research/frameworks/htm/temporal_memory/sequence_memory_apical_tiebreak.py (hunk 2 above)
Comment (abhi-iyer) [reply src/nupic/research/frameworks/htm/temporal_memory/sequence_memory_apical_tiebreak.py:76]: Will incorporate

**Review by lscheinkman — APPROVED** [review]

File: src/nupic/research/frameworks/htm/temporal_memory/sequence_memory_apical_tiebreak.py (hunk 3 above)
Comment (lscheinkman) [reply src/nupic/research/frameworks/htm/temporal_memory/sequence_memory_apical_tiebreak.py:107]: The current docstring style is the one recommended by https://peps.python.org/pep-0287/
I am open to change any standard docstring format that is well documented and supported by sphinx and readthedocs. See https://nupicresearch.readthedocs.io/en/latest/

File: src/nupic/research/frameworks/htm/temporal_memory/pair_memory_apical_tiebreak.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+device = "cuda" if torch.cuda.is_available() else "cpu"
+
+
+class PairMemoryApicalTiebreak(TemporalMemoryApicalTiebreak):
+    """
+    pair memory with apical tiebreak.
```
[the remark this answers is not in the document]
Comment (abhi-iyer) [reply src/nupic/research/frameworks/htm/temporal_memory/pair_memory_apical_tiebreak.py:34]: will elaborate

File: src/nupic/research/frameworks/htm/temporal_memory/sequence_memory_apical_tiebreak.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+real_type = torch.float32
+int_type = torch.int64
+
+device = "cuda" if torch.cuda.is_available() else "cpu"
+
+
+class SequenceMemoryApicalTiebreak(TemporalMemoryApicalTiebreak):
+    """
+    sequence memory with apical tiebreak, built on temporal memory with apical tiebreak.
```
[the remark this answers is not in the document]
Comment (abhi-iyer) [reply src/nupic/research/frameworks/htm/temporal_memory/sequence_memory_apical_tiebreak.py:34]: I'll elaborate on the comment -- Marcus has two separate classes like this as well (SequenceMemory and PairMemory)

File: src/nupic/research/frameworks/htm/temporal_memory/temporal_memory_apical_tiebreak.py [hunk 6]
```diff
@@ -0,0 +1,1324 @@
...
+
+real_type = torch.float32
+int_type = torch.int64
+
+device = "cuda" if torch.cuda.is_available() else "cpu"
+
+
+class TemporalMemoryApicalTiebreak():
+    """
+    generalized Temporal Memory (TM) with apical dendrites that add a "tiebreak".
```
[the remark this answers is not in the document]
Comment (abhi-iyer) [reply src/nupic/research/frameworks/htm/temporal_memory/temporal_memory_apical_tiebreak.py:34]: I'm copying comments from here: https://github.com/numenta/nupic.research/blob/master/packages/columns/src/nupic/research/frameworks/columns/apical_tiebreak_temporal_memory.py, which is the C++ implementation

[the remark this answers is not in the document]
Comment (abhi-iyer) [reply src/nupic/research/frameworks/htm/temporal_memory/temporal_memory_apical_tiebreak.py:34]: generalized is later explaining in the same block comment: 

```
the class is generalized in two ways:
        - exposes two main methods `depolarize_cells` and `activate_cells`.
        - TM is unaware of whether `basal_input` or `apical_input` are from
          internal/external cells -- caller knows what these cells numbers mean
        - TM does not specify when a "timestep" begins and ends. callers or subclasses
          can introduce the notion of a timestep.
```

**Review by subutai — APPROVED** [review]

File: src/nupic/research/frameworks/htm/temporal_memory/sequence_memory_apical_tiebreak.py (hunk 3 above)
Comment (lucasosouza) [reply src/nupic/research/frameworks/htm/temporal_memory/sequence_memory_apical_tiebreak.py:133]: Might have been misunderstood here. I am not suggesting we change docstring format. In this comment, I was suggesting we change the docstrings in this PR to match the current standard used in the rest of the repo.

**Pull request merged**